File: src/nuspace/web/refs/pages/ops.py
"""The Pages op table: one wire path per op, one body per path.

The browser notifies ``<pages ref>.ops.<name>`` and the payload carries
that op's arguments and nothing else. There is no ``op`` key, because the
path already answered that question.

    op               payload
    --               --
    page.select      {path}
    page.create      {parent_path, title}
    page.rename      {path, title}
    page.delete      {path}
    block.create     {page_path, tpl, content, after}
    block.update     {page_path, block_id, content}
    block.delete     {page_path, block_ids}
    block.split      {page_path, block_id, head, tail, insert}
    block.merge      {page_path, block_id, into_id, content}
    block.reorder    {page_path, order}
    block.restart    {block_id}

``insert`` is ``{tpl, content} | None``. ``content`` is whatever the tpl
takes: python source for a ``program``, markdown for a ``text``. Nothing
here branches on which -- ``BlockOps.set_content`` asks the registry.

``target`` is the Ref each op's effect lands on. Everything that changes
the document names ``Space.pages``; ``page.select`` and ``block.restart``
change what this connection runs and shows, so they name the surface
ref. A Command's mutation slot is an address, and these are the honest
ones.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from nuspace.web.refs.common import Ops
from nuspace.web.refs.pages.interactions import BlockOps, PageOps


if TYPE_CHECKING:
    import nu
    from nuspace.web.refs.pages.ref import PagesRef
    from nuspace.web.refs.pages.view import View


__all__ = ["dispatch"]


def dispatch(base: str, ref: PagesRef, view: View) -> nu.Nu:
    """Build the reactive handler for every op, composed in parallel."""
    ops = Ops(base)
    pages = PageOps(view)
    blocks = BlockOps(view)
    kv = view.root.pages

    handlers = [
        ops.on("page.select", ref, pages.select),
        ops.on("page.create", kv, pages.create),
        ops.on("page.rename", kv, pages.rename),
        ops.on("page.delete", kv, pages.delete),
        ops.on("block.create", kv, blocks.create),
        ops.on("block.update", kv, blocks.update),
        ops.on("block.delete", kv, blocks.delete),
        ops.on("block.split", kv, blocks.split),
        ops.on("block.merge", kv, blocks.merge),
        ops.on("block.reorder", kv, blocks.reorder),
        ops.on("block.restart", ref, blocks.restart),
        ops.stray(ref),
    ]
    composed = handlers[0]
    for handler in handlers[1:]:
        composed = composed | handler
    return composed

File: src/nuspace/web/ui/src/refs/pages/ops.ts
// The Pages op table, browser side. Mirrors `nuspace/web/refs/pages.py`.
//
// One ref per op: a notify goes to `<pages ref>.ops.<name>` and the payload
// carries that op's arguments and nothing else. There is no `op` key, because
// the path already answered that question, and the server binds one arm per
// path rather than branching on a string.
//
// `Ops` is what keeps the two sides in step: the keys are the op names the
// python ref subscribes to, and each value is exactly the field set the
// matching `nuspace.pages.ops` call reads. Add an interaction on one side and
// this file is where the other side fails to compile.
//
// ## Ids are minted here
//
// `page.create` and `section.create` carry the id of the thing being made.
// That makes a create a pure function of its event on the server, and lets a
// click route to what it just made without a round trip to learn its name.
// Use `mintId` from ./types; do not invent a second scheme.
//
// ## What is NOT here
//
// Typing in a text block. A text block's ref is wired straight to kv by the
// block's own program, so a keystroke is a notify on that ref and never
// touches the page document.
//
// Split and merge. Both are compositions of the ops below -- a split is an
// update plus a create at the next index, a merge is an update plus a delete
// -- and the canvas composes them rather than the server growing a verb for
// each editor gesture. The two halves are independent writes to different
// sections, so it does not matter which arm runs first.
//
// Restart. There is no liveness pillar to restart into yet.

import type { BlockTpl } from "./types";

/** Every op the Pages surface accepts, with its argument shape. */
export type Ops = {
	/** No kv write: the answer is a `set_page` plus a `set_status`. */
	"page.select": { page_id: string };
	"page.create": { page_id: string; parent_id: string; title: string };
	"page.rename": { page_id: string; title: string };
	/** Takes the whole subtree with it. */
	"page.delete": { page_id: string };
	"page.move": { page_id: string; parent_id: string; index: number };
	/** One parent's children, in the order they should now be in. */
	"page.reorder": { parent_id: string; page_ids: string[] };
	"section.create": {
		page_id: string;
		section_id: string;
		name: string;
		tpl: BlockTpl;
		/** The Nu program. For a templated tpl this is the shipped starter,
		 *  which arrives in the ref's mount props -- never written here. */
		source: string;
		index: number;
	};
	/** Replaces the stored program. Nothing else about the section moves. */
	"section.update": { page_id: string; section_id: string; source: string };
	"section.delete": { page_id: string; section_id: string };
	/** Keeps the section's id, and so its kv namespace, on the new page. */
	"section.move": {
		page_id: string;
		section_id: string;
		to_page_id: string;
		index: number;
	};
	"section.reorder": { page_id: string; section_ids: string[] };
};

/** Send one op. Threaded down from `pages.tsx` into the rail and the canvas. */
export type Notify = <K extends keyof Ops>(op: K, args: Ops[K]) => void;
